core/templates/pages/admin-page/misc-tab/admin-misc-tab.component.ts

```diff
@@ -60,13 +76,138 @@ export class AdminMiscTabComponent {
+    if (this.areUserGroupsUpdated()) {
+      if (
+        !this.windowRef.nativeWindow.confirm(
+          'This will revert all changes you made. Are you sure?'
+        )
+      ) {
+        return;
+      }
+      this.userGroupsToUsers = cloneDeep(this.userGroupToUsersMapBackup);
+    }
+  }
+
+  areUserGroupsUpdated(): boolean {
+    return !isEqual(this.userGroupsToUsers, this.userGroupToUsersMapBackup);
+  }
+
+  getUserGroups(): string[] {
+    let userGroups: string[] = [];
+    for (let key in this.userGroupsToUsers) {
+      userGroups.push(key);
+    }
+    return userGroups;
+  }
+
+  addUser(event: {value: string}): void {
+    const value = (event.value || '').trim();
+    if (!value || value === '') {
+      return;
+    }
+
+    if (this.userGroupsToUsers[this.selectedUserGroup].includes(value)) {
+      this.alertsService.addWarning(
+        `The user ${value} already exists in the user ` +
+          `group ${this.selectedUserGroup}.`
+      );
+      return;
+    }
+    this.userGroupsToUsers[this.selectedUserGroup].push(value);
+    this.userInput.nativeElement.value = '';
+  }
+
+  removeUser(username: string): void {
+    let usersOfSelectedUserGroup: string[] =
+      this.userGroupsToUsers[this.selectedUserGroup];
+    this.userGroupsToUsers[this.selectedUserGroup] =
+      usersOfSelectedUserGroup.filter(obj => obj !== username);
+  }
+
+  removeUserGroup(): void {
+    delete this.userGroupsToUsers[this.selectedUserGroup];
+    this.selectedUserGroup = '';
+  }
+
+  addUserGroup(): void {
+    if (this.newUserGroupName.trim() in this.userGroupsToUsers) {
+      this.alertsService.addWarning(
+        `The user group ${this.newUserGroupName} already exists.`
+      );
+      return;
+    }
+
+    if (this.newUserGroupName.trim() !== '') {
+      this.userGroupsToUsers[this.newUserGroupName.trim()] = [];
+      this.selectedUserGroup = this.newUserGroupName.trim();
+      this.newUserGroupName = '';
+    }
+  }
+
+  getUsersForSelectedUserGroup(): string[] {
+    if (this.selectedUserGroup in this.userGroupsToUsers) {
+      return this.userGroupsToUsers[this.selectedUserGroup];
+    }
+    return [];
+  }
+
+  updateUserGroups(): void {
+    if (!this.areUserGroupsUpdated()) {
+      return;
+    }
+
+    if (
+      this.adminTaskManagerService.isTaskRunning() ||
+      !this.windowRef.nativeWindow.confirm(this.irreversibleActionMessage)
+    ) {
+      return;
+    }
+
+    this.setStatusMessage.emit('Updating UserGroups...');
+
+    this.adminTaskManagerService.startTask();
+    this.adminBackendApiService
+      .updateUserGroupsAsync(this.userGroupsToUsers)
+      .then(
+        () => {
+          this.setStatusMessage.emit('UserGroups successfully updated.');
+          this.userGroupToUsersMapBackup = cloneDeep(this.userGroupsToUsers);
+          this.adminTaskManagerService.finishTask();
+        },
+        errorResponse => {
+          this.setStatusMessage.emit('Server error: ' + errorResponse);
```

Comment:
Use string interpolation instead of concatenation.

Reply:
Updated